site/content/protocols/messagepickup/4.0/readme.md

```diff
@@ -0,0 +1,388 @@
+---
+title: Message Pickup
+publisher: JamesKEbert
+license: MIT
+piuri: https://didcomm.org/message-pickup/4.0
+status: Production
+summary: A protocol to facilitate an agent picking up messages held at a mediator.
+tags: []
+authors:
+  - name: Sam Curren
+    email: telegramsam@gmail.com
+  - name: James Ebert
+    email: james@jamesebert.dev
+
+---
+
+## Summary
+A protocol to facilitate a _Recipient_ agent picking up messages held at a _Mediator_. This protocol is likely to be used in tandem with the [cooridinate-mediation protocol](https://didcomm.org/coordinate-mediation/3.0/).
+
+## Motivation
+This protocol is needed to facilitate retrieval of messages from a mediator in an explicit manner. Additionally, this protocol provides behavior for initiating live delivery of messages, which is crucial for good user experience for agents operating on mobile devices.
+
+Motivation for v4 of this protocol primarily stems from ambiguity in the [pickup v2 protocol](https://github.com/hyperledger/aries-rfcs/tree/main/features/0685-pickup-v2) and [messagepickup v3 protocol](https://didcomm.org/messagepickup/3.0/) as to whether `delivery` and `messages-received` messages must be used while using live mode. 
+
+## Version Change Log
+
+### 4.0
+ - Clarifies that `delivery` and `messages-received` messages MUST be used while using _Live Mode_ to ensure delivery of all messages.
+ - Adjusts handling of `delivery-request` messages when no messages are queued to be delivered. Rather than sending a status message indicating no messages are queued, an empty delivery message is sent. This was chosen to reduce complexity of the protocol.
+ - Adds DIDComm v1 message format (pickup v3 only contained DIDComm v2).
+    - In pickup v4 while using DIDComm v1, every `recipient_did` must be a [`did:key` reference](https://github.com/hyperledger/aries-rfcs/tree/main/features/0360-use-did-key).
+ - Adjustments to the protocol's use of thread ids
+    - Added thread id to `status` messages, as status messages are always in response to an original message from the `recipient`.
+    - Made the thread id (thid) on `delivery` messages optional, as messages received in _Live Mode_ do not have an associated `delivery-request` message.
+ - Removed return route all from `messages-received` messages, as it is not required.
+ - Updates protocol-name to `message-pickup`.
+
+## Roles
+There are two roles in this protocol: 
+
+- `mediator`: The agent that has messages waiting for pickup by the `recipient`.
```

Comment:
> Some time ago I had an use case where mobile agents without mediator would get credentials from an issuer: when connecting through regular HTTP, we were using Message Pickup in polling mode to check for credential issuance and other messages until the flow was finished. Do you think this is not an intended usage for this protocol?

My current answer is that I think I would consider that to be outside the scope/intentions of this protocol. Given the already fairly involved nature of this protocol, my initial response is that I'd rather have a separate protocol dedicated to that particular use case so it can be more specific in its definitions and clarifications (especially given the troublesome, hard to troubleshoot issues observed with mobile message pickup over the years).

Also some context there according to my memory -- Pickup v2 was a completely fresh draft of the protocol that did not draw heavily from the v1 version (it was more based off of the implicit mechanisms being used with ACA-Py), but we thought it would be ideal to replace the initial version.